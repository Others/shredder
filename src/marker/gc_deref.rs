use std::prelude::v1::*;

/// A marker trait that marks that this data can be stored in a `DerefGc`
///
/// `T` can be `GcDeref` only if it is deeply immutable through a `&T`. This is because it's
/// problematic to move `Gc`s around from inside other `Gc`s without a `GcGuard`.
///
/// This would seem to make `DerefGc`/`GcDeref` very hard to use. Luckily there is one exception! A
/// `GcDeref` can contain a `Gc`, even if that `Gc` contains data that is not deeply immutable
/// through`&T`!
///
/// You can get the `Scan` derive to automatically generate your `GcDeref` implementation for you.
/// To do this, you should supply the `#[shredder(can_deref)]` annotation. For example:
/// ```
/// use shredder::Scan;
///
/// #[derive(Scan)]
/// #[shredder(can_deref)]
/// struct ToPutInADerefGc {
///     random_data: u32
/// }
/// // `ToPutInADerefGc` will automagically get a `GcDeref` implementation
/// ```
pub unsafe trait GcDeref: Sync {}
