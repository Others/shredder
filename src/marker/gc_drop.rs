/// A marker trait that the destructor of this data can be safely run in the background thread
///
/// Basically it asserts three things
/// 1) Any thread can drop this data (It is `Send`, or `!Send` purely because it contains a `Gc`, or
/// it is `!Send` but you know that any thread dropping this data is safe.)
/// 2) This data does not own a `AtomicGc` or `DerefGc`
/// 3) This data is `'static`, or you can guarantee that it's safe to drop it after its lifetime
/// has ended.
///
/// This trait is structural, although I'm sure with some crazy unsafe code you could break that
/// assumption.
///
/// The `Scan` derive will automatically generate a `GcDrop` impl for you. For example:
/// ```
/// use shredder::Scan;
///
/// // SomeGcDropType will implement `GcDrop` as well as `Scan` and `GcSafe`
/// #[derive(Scan)]
/// struct SomeGcDropType {
///     random_data: u32
/// }
/// ```
///
/// If this is causing you problems you can turn it off with the `#[shredder(cant_drop)]`
/// annotation. For example:
/// ```
/// use shredder::atomic::AtomicGc;
/// use shredder::Scan;
///
/// // CantImplGcDrop will not get a `GcDrop` implementation
/// #[derive(Scan)]
/// #[shredder(cant_drop)] // <- This allows this code to compile
/// struct CantImplGcDrop {
///     some_non_gc_drop_type: AtomicGc<u32>
/// }
/// ```
pub unsafe trait GcDrop {}
