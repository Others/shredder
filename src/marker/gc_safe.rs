use std::hash::{Hash, Hasher};
use std::ops::{Deref, DerefMut};

/// A marker trait that marks that data can be scanned in the background by the garbage collector.
///
/// Specifically, the requirement for implementing this trait is:
///  1) It's okay for any thread to call `scan` on this data, as long as it has exclusive access to
///     the data.
///  2) It is safe for `scan` to be called on this data, even if the lifetime of this data has
///     ended.
///
/// Importantly if a type is `Send + 'static`, then it is always `GcSafe`
///
/// NOTE: `GcSafe` cannot simply be `Send + 'static`, since `Gc` is always `GcSafe` but sometimes is
/// not `Send` or `'static`
pub unsafe trait GcSafe {}

/// `GcSafeWrapper` wraps a `Send + 'static` datatype to make it `GcSafe`
///
/// Usually you'd just want to implement `GcSafe` directly, but this is useful situationally if
/// you need to use someone else's type which does not implement `GcSafe`.
///
/// TODO: Remove this once overlapping marker traits are stabilized
pub struct GcSafeWrapper<T> {
    /// The wrapped value
    pub v: T,
}

unsafe impl<T> GcSafe for GcSafeWrapper<T> {}

impl<T> GcSafeWrapper<T> {
    /// Create a new `GcSafeWrapper` storing `v`, assuming `v: Send + 'static`
    ///
    /// Safe since all `Send + 'static` data is `GcSafe` automatically
    pub fn new(v: T) -> Self
    where
        T: Send + 'static,
    {
        Self { v }
    }

    /// Create a new `GcSafeWrapper` storing `v`, ignoring whether it's `Send + 'static`
    ///
    /// # Safety
    /// Only safe to use if you can guarantee the value passed in fulfills the requirements to
    /// implement `GcSafe`.
    pub unsafe fn new_unsafe(v: T) -> Self {
        Self { v }
    }

    /// `take` the value out of this wrapper
    pub fn take(self) -> T {
        self.v
    }
}

impl<T: Send + Clone> Clone for GcSafeWrapper<T> {
    fn clone(&self) -> Self {
        Self { v: self.v.clone() }
    }
}

impl<T: Send + Copy> Copy for GcSafeWrapper<T> {}

impl<T: Send + Default> Default for GcSafeWrapper<T> {
    fn default() -> Self {
        Self { v: T::default() }
    }
}

impl<T: Send> Deref for GcSafeWrapper<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.v
    }
}

impl<T: Send> DerefMut for GcSafeWrapper<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.v
    }
}

impl<T: Send + Hash> Hash for GcSafeWrapper<T> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.v.hash(state);
    }
}

#[allow(clippy::partialeq_ne_impl)]
impl<T: Send + PartialEq> PartialEq for GcSafeWrapper<T> {
    fn eq(&self, other: &Self) -> bool {
        self.v.eq(other)
    }

    fn ne(&self, other: &Self) -> bool {
        self.v.ne(other)
    }
}
impl<T: Send + Eq> Eq for GcSafeWrapper<T> {}
