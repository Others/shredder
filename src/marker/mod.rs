mod gc_deref;
mod gc_drop;
mod gc_safe;

pub use gc_deref::GcDeref;

pub use gc_drop::GcDrop;

pub use gc_safe::GcSafe;
pub use gc_safe::GcSafeWrapper;
