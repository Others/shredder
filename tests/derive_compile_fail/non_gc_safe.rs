extern crate shredder;

use shredder::Scan;
use shredder::marker::GcDrop;

struct NotGcSafe {}
unsafe impl GcDrop for NotGcSafe {}

#[derive(Scan)]
struct Test0 {
    #[shredder(skip_scan)]
    field: NotGcSafe
}

fn main() {}
