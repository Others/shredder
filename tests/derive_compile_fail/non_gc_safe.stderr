error[E0277]: the trait bound `NotGcSafe: shredder::marker::GcSafe` is not satisfied
  --> $DIR/non_gc_safe.rs:9:10
   |
9  | #[derive(Scan)]
   |          ^^^^ the trait `shredder::marker::GcSafe` is not implemented for `NotGcSafe`
   |
  ::: $WORKSPACE/src/plumbing.rs
   |
   | pub fn check_gc_safe<T: GcSafe>(_: &T) {}
   |                         ------ required by this bound in `shredder::plumbing::check_gc_safe`
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
